### DISCUSSION TITLE
More Codecs

### DISCUSSION THREAD

**Discussion by ZavierQuixote** [TT-ReBORN/Georgia-ReBORN#170] [Ideas]
I created 10 more images for codecs to add to those already inclded in the Georgia Reborn theme.  Also included are updated ALAC and Ogg images with the newest logo and the Vorbis fish, respectively.  Also included in the attached archive is an updated gr-main-functions.js to make the codecs appear on the Details tab in the player.  PNGs go in the foobar2000\profile\georgia-reborn\images\codec folder and JS in the foobar2000\profile\georgia-reborn\scripts\Base folder.  Additional foobar2000 components are needed for these additional codecs; excepting ALAC, Vorbis and Speex.
[codecs.zip](https://github.com/TT-ReBORN/Georgia-ReBORN/files/14158600/codecs.zip)

**Comment by TT-ReBORN** [comment]
Hi @ZavierQuixote,

thanks a lot for your work! I will add them to the existing codecs once I have the time for it.
Just to be sure, you checked every each of them to display correctly with your added code right?

-TT

**Reply by ZavierQuixote** [reply]
I'd checked each new codec to make sure it works yesterday.  Tonight I checked to make sure nothing I changed in the javascript file adversely affected the pre-existing codecs' functionality.  I found that codecs used in a .wav file don't register with a .png.  I'm currently working on a fix for that...which will probably entail a few more codecs.

**Reply by ZavierQuixote** [reply]
Attached is an archive containing all codecs including some new Audio Compression Manager codecs contained in .wav files (including MP3, AAC, AC3 & DTS) along with a cleaned-up JS file.  I also made a new PNG for the MP2 file type.  
[codec.zip](https://github.com/TT-ReBORN/Georgia-ReBORN/files/14175769/codec.zip)
The LA file type has no 64-bit plugin, unless someone makes one before Georgia Reborn's required plugins all have 64-bit versions; so that codec will probably need removing then.

**Reply by ZavierQuixote** [reply]
Come to think of it, the Shorten foobar plugin is also only available in 32-bit.  Then again, both it and LA are quite antiquated.

**Reply by TT-ReBORN** [reply]
Thanks, I am in the process of finishing refactoring the whole theme.
I have rewritten a lot of functions, the code base now has much cleaner structure and also faster performance.

Can you add the new codecs into this new method, i.e the `codecFormat` object? I would appreciate it:

```
	loadCodecLogo() {
		const codec = $('$lower($if2(%codec%,$ext(%path%)))');
		const format = $('$lower($ext(%path%))', fb.GetNowPlaying());
		const lightBg = new Color(grCol.detailsText).brightness < 140;
		const bw = lightBg ? 'black' : 'white';

		const codecFormat = {
			'aac':  'aac',
			'ac3':  'ac3', 'atsc a/52': 'ac3', 'e-ac3': 'ac3',
			'aiff': 'pcm-aiff',
			'alac': 'alac',
			'ape':  'ape', 'monkey\'s audio': 'ape',
			'dsd':  'dsd',
			'dst':  'dsd-sacd',
			'dts':  'dts', 'dca (dts coherent acoustics)': 'dts',
			'dxd':  'dxd',
			'flac': 'flac',
			'mlp':  'mlp',
			'mp3':  'mp3',
			'mpc':  'musepack', 'musepack': 'musepack',
			'ogg':  'ogg', 'vorbis': 'ogg',
			'opus': 'opus',
			'pcm':  format === 'aiff' ? 'pcm-aiff' : ['w64', 'wav'].includes(format) ? 'pcm-wav' : 'pcm',
			'tak': 'tak',
			'tta': 'tta', 'true audio': 'tta',
			'wav': 'pcm-wav',
			'w64': 'pcm-wav',
			'wma': 'wma',
			'wv':  'wavpack', 'wavpack': 'wavpack'
		};

		const codecName = codecFormat[codec] || codecFormat[format];
		const codecLogoPath = (codecName) => `${grPaths.imagesPath}codec\\${codecName}-${bw}.png`;

		if (codecName) {
			grImg.codecLogo = gdi.Image(codecLogoPath(codecName));
		}
		// Handle special cases
		if (codec.startsWith('dsd')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dsd'));
		} else if (codec.startsWith('dxd')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dxd'));
		} else if (codec.startsWith('dst')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dst'));
		}
	}
```

**Reply by ZavierQuixote** [reply]
Here you go (at least, I believe this would be correct):

**Reply by ZavierQuixote** [reply]
loadCodecLogo() {
	const codec = $('$lower($if2(%codec%,$ext(%path%)))');
	const format = $('$lower($ext(%path%))', fb.GetNowPlaying());
	const lightBg = new Color(col.detailsText).brightness < 140;
	const bw = lightBg ? 'black' : 'white';

		const codecFormat = {
			'aac':  'aac', 'aac acm codec': 'aac',
			'ac3':  'ac3', 'atsc a/52': 'ac3', 'e-ac3': 'ac3',
			'aiff': 'pcm-aiff',
			'alac': 'alac',
			'alaw': 'alaw', 'ccitt a-law': 'alaw',
			'amr': 'amr',
			'ape':  'ape', 'monkey\'s audio': 'ape',
			'caf': 'caf',
			'dsd':  'dsd',
			'dst':  'dsd-sacd',
			'dts':  'dts', 'dca (dts coherent acoustics)': 'dts',
			'dxd':  'dxd',
			'flac': 'flac',
			'gsm': 'gsm', 'gsm 6.10': 'gsm',
			'imaadpcm': 'imaadpcm', 'ima adpcm': 'imaadpcm',
			'la':  'la',
			'mid': 'mid',
			'mlp':  'mlp',
			'mod': 'mod',
			'mp2':  'mp2',
			'mp3':  'mp3', 'mpeg layer-3': 'mp3',
			'mpc':  'musepack', 'musepack': 'musepack',
			'msadpcm': 'msadpcm', 'microsoft adpcm': 'msadpcm',
			'ofr':  'ofr', 'optimfrog': 'ofr',
			'ogg':  'ogg', 'vorbis': 'ogg',
			'opus': 'opus',
			'pcm':  format === 'aiff' ? 'pcm-aiff' : ['w64', 'wav'].includes(format) ? 'pcm-wav' : 'pcm',
			'qoa': 'qoa',
			'shn': 'shn', 'shorten': 'shn',
			'spx': 'spx', 'speex': 'spx',
			'tak': 'tak',
			'tta': 'tta', 'true audio': 'tta',
			'ulaw': 'ulaw', 'ccitt u-law': 'ulaw',
			'usac': 'usac',
			'wav': 'pcm-wav',
			'w64': 'pcm-wav',
			'wma': 'wma',
			'wv':  'wavpack', 'wavpack': 'wavpack'
		};

		const codecName = codecFormat[codec] || codecFormat[format];
		const codecLogoPath = (codecName) => `${grPaths.imagesPath}codec\\${codecName}-${bw}.png`;

		if (codecName) {
			grImg.codecLogo = gdi.Image(codecLogoPath(codecName));
		}
		// Handle special cases
		if (codec.startsWith('dsd')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dsd'));
		} else if (codec.startsWith('dxd')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dxd'));
		} else if (codec.startsWith('dst')) {
			grImg.codecLogo = gdi.Image(codecLogoPath('dst'));
		}
	}

**Reply by ZavierQuixote** [reply]
By the way, I compressed all the pngs from your foobar skin with oxipng and zopfli to achieve the smallest file sizes losslessly.  In all, it shaves about 1.2 MB from the total space used.  If you'd like copies of the smaller pngs, let me know and I can upload them for you.

**Reply by TT-ReBORN** [reply]
Thanks for the work. 

Actually all Georgia-ReBORN images were already processed and optimized by [pngquant](https://github.com/kornelski/pngquant) except the disc art placeholders. 
I've tried 6 different png optimizers with various settings and they all failed. They produced jagged edges/anti-aliasing issues, 
when the disc art placeholders are spinning, this can be seen especially on 4K displays. 

So only the disc art placeholders are original unprocessed png images, all other images are already optimized. 
I will optimize your new codec logos.

-TT

**Comment by ZavierQuixote** [comment]
I did use pngquant on my codec logos before using oxipng on them; as I noticed the others from the theme were in 8-bit color depth.

**Reply by TT-ReBORN** [reply]
So you are saying that when using oxipng and zopfli on the already optimized pngquant images,
it will compress and save additionally 1.2 MB in total?

**Reply by ZavierQuixote** [reply]
The unzipped "images" folder from Georgia-ReBORN_v3.0-RC2.zip is 15.8 MB.  With my added codec pngs, that total folder size after using oxipng's settings of "opt max", "strip safe" and "zopfli" yields a total 14.2 MB; so it's actually 1.6 MB smaller.  (1.2 was a conservative estimate, as I wasn't at my computer at the time.)  If you want to test the ones I edited from the "discart" subfolder with the rotating images in 4K (I don't have a monitor bigger than FHD to try it on), I can upload/send those to you in an archive file.

**Reply by TT-ReBORN** [reply]
Sure, you can send me the zip and I'll take a look once I have more time, thanks!

**Reply by ZavierQuixote** [reply]
(The 1400 x 1400 pixel images do take a while to process.)

**Reply by ZavierQuixote** [reply]
Here are the amended contents of the discart folder.
[discart.zip](https://github.com/TT-ReBORN/Georgia-ReBORN/files/14209307/discart.zip)

**Reply by ZavierQuixote** [reply]
...and for kicks and giggles, the flags folder
[flags.zip](https://github.com/TT-ReBORN/Georgia-ReBORN/files/14209325/flags.zip)

**Reply by ZavierQuixote** [reply]
...and the misc folder.
[misc.zip](https://github.com/TT-ReBORN/Georgia-ReBORN/files/14209334/misc.zip)

**Comment by ZavierQuixote** [comment]
Sorry this is so long in coming, but I finally got Gigapixel AI and updated the two optimfrog pngs for the codecs folder with improved icons.
<img width="215" height="60" alt="ofr-black" src="https://github.com/user-attachments/assets/b0940bf2-fa69-4216-843c-5ccf9ee46a4b" />
<img width="215" height="60" alt="ofr-white" src="https://github.com/user-attachments/assets/c5d491c6-dda1-403d-ba01-fccf76b7f301" />

**Reply by TT-ReBORN** [reply]
Hi @ZavierQuixote,

this is definitely a good improvement over the original ones when viewed in 100% original size. 
I did not try Gigapixel for a long time, but it was my go-to reference upscaler. Is it still the best vs the competition? 
Also did you try various settings and modes to soften up the black contours on the frog?

-TT
